driver-core/src/main/com/mongodb/internal/connection/UpdateMessage.java

```diff
@@ -55,10 +56,19 @@ protected EncodingMetadata encodeMessageBodyWithMetadata(final BsonOutput bsonOu
 
         addDocument(updateRequest.getFilter(), bsonOutput, new NoOpFieldNameValidator());
         if (updateRequest.getType() == REPLACE) {
-            addDocument(updateRequest.getUpdate(), bsonOutput, new CollectibleDocumentFieldNameValidator());
+            addDocument(updateRequest.getUpdate().asDocument(), bsonOutput, new CollectibleDocumentFieldNameValidator());
         } else {
             int bufferPosition = bsonOutput.getPosition();
-            addDocument(updateRequest.getUpdate(), bsonOutput, new UpdateFieldNameValidator());
+            BsonValue update = updateRequest.getUpdate();
+            if (update.isDocument()) {
+                addDocument(update.asDocument(), bsonOutput, new UpdateFieldNameValidator());
+            } else if (update.isArray()) {
```

Comment:
I think we can ditch this code path.  The reason is because this is the code path used for OP_UPDATE wire protocol message, and we only exercise this path when both of these conditions are true:

* Server version is < 3.6 (after which we use OP_MSG)
* Write concern is unacknowledged

Since update pipelines are a new feature, in practice this code would never work were the message to reach the server.  So it's sufficient to just throw the IllegalArgumentException.

Reply:
I have remove the check for `isArray()` and throw the `IllegalArgumentException`.